File: backend/python/app/services/interfaces/simple_entity_service.py
from abc import ABC, abstractmethod


class ISimpleEntityService(ABC):
    """
    A class to handle CRUD functionality for simple entities
    """

    @abstractmethod
    def get_entities(self):
        """Return a list of all simple entities

        :return: A list of dictionaries from SimpleEntity objects
        :rtype: list of dictionaries
        """
        pass

    @abstractmethod
    def get_entity(self, id):
        """Return a dictionary from the SimpleEntity object based on id

        :param id: SimpleEntity id
        :return: dictionary of SimpleEntity object
        :rtype: dictionary
        :raises Exception: id retrieval fails
        """
        pass

    @abstractmethod
    def create_entity(self, entity):
        """Create a new SimpleEntity object

        :param entity: dictionary of simple entity fields
        :return: dictionary of SimpleEntity object
        :rtype: dictionary
        :raises Exception: if simple entity fields are invalid
        """
        pass

    @abstractmethod
    def update_entity(self, id, entity):
        """Update existing simple entity

        :param entity: dictionary of simple entity fields
        :param id: SimpleEntity id
        :return: dictionary of SimpleEntity object
        :rtype: dictionary
        """
        pass

    @abstractmethod
    def delete_entity(self, id):
        """Delete existing simple entity

        :param id: SimpleEntity id
        :return: id of the SimpleEntity deleted
        :rtype: integer
        """
        pass

File: backend/typescript/services/interfaces/simpleEntityService.ts
export interface SimpleEntityRequestDTO {
  stringField: string;
  intField: number;
  enumField: string;
  stringArrayField: string[];
  boolField: boolean;
}

export interface SimpleEntityResponseDTO {
  id: string;
  stringField: string;
  intField: number;
  enumField: string;
  stringArrayField: string[];
  boolField: boolean;
}

export interface ISimpleEntityService {
  /**
   * retrieve the SimpleEntity with the given id
   * @param id SimpleEntity id
   * @returns requested SimpleEntity
   * @throws Error if retrieval fails
   */
  getEntity(id: string): Promise<SimpleEntityResponseDTO>;

  /**
   * retrieve all SimpleEntities
   * @param
   * @returns returns array of SimpleEntities
   * @throws Error if retrieval fails
   */
  getEntities(): Promise<SimpleEntityResponseDTO[]>;

  /**
   * create a SimpleEntity with the fields given in the DTO, return created SimpleEntity
   * @param entity new SimpleEntity
   * @returns the created SimpleEntity
   * @throws Error if creation fails
   */
  createEntity(
    entity: SimpleEntityRequestDTO,
  ): Promise<SimpleEntityResponseDTO>;

  /**
   * update the SimpleEntity with the given id with fields in the DTO, return updated SimpleEntity
   * @param id SimpleEntity id
   * @param entity Updated SimpleEntity
   * @returns the updated SimpleEntity
   * @throws Error if update fails
   */
  updateEntity(
    id: string,
    entity: SimpleEntityRequestDTO,
  ): Promise<SimpleEntityResponseDTO | null>;

  /**
   * delete the SimpleEntity with the given id
   * @param id SimpleEntity id
   * @returns id of the SimpleEntity deleted
   * @throws Error if deletion fails
   */
  deleteEntity(id: string): Promise<string>;
}
